reference/cd-codefresh.md

```diff
@@ -0,0 +1,118 @@
+
+To use Pulumi within Codefresh, there are a few environment variables you'll need to set for each
+build.
+
+The first is `PULUMI_ACCESS_TOKEN`, which is required to authenticate with pulumi.com in order to
+perform the preview or update. You can create a new Pulumi access token specifically for your
+CI/CD job on your [Pulumi Account page](https://app.pulumi.com/account/tokens).
+
+You can either add the token on the pipeline itself as a variable, or store it globally using [Codefresh shared configuration](https://codefresh.io/docs/docs/configure-ci-cd-pipeline/shared-configuration/).
+
+![Pulumi token in Codefresh](/images/reference/codefresh/codefresh-pulumi-token.png)
+
+Next, you will also need to set environment variables specific to your cloud resource provider.
+For example, if your stack is managing resources on AWS, `AWS_ACCESS_KEY_ID` and
+`AWS_SECRET_ACCESS_KEY`.
+
+
+## Codefresh CI/CD pipeline with Pulumi
+
+In your pipeline you need at least two steps.
+
+1. A step that downloads all Pulumi dependencies according to your programming language
+1. A step that performs the actual deployment 
+
+In all cases you use a [Codefresh freestyle step](https://codefresh.io/docs/docs/codefresh-yaml/steps/freestyle/) with the Pulumi Docker image.
+
+```yaml
+  RunMyPulumiStep:
+    title: Running Pulumi inside Codefresh
+    image: pulumi/pulumi
+    commands: 
+      - pulumi stack --non-interactive
+```
+
+Pulumi will search for a login token named `PULUMI_ACCESS_TOKEN` so assuming that you have setup this variable in the pipeline, Pulumi will not ask for interactive login.
+
+For Kubernetes based deployments, there is no other setup needed. Codefresh is setting up automatically a `kubeconfig` in all freestyle steps which is the same mechanism used by `kubectl` and Pulumi to access Kubernetes clusters. You only need to select your cluster if you have more than one, using a `kubectl config use-context` command prior to running `pulumi`.
+
+For other non-Kubernetes deployments, you need to add additional environment variables in your pipeline that will be used by Pulumi for accessing your cloud provider.
+
+Here is a full example:
+
+ `codefresh.yml`
+{% highlight yaml %}
+{% raw %}
+version: '1.0'
+stages:
+  - prepare
+  - build
+  - deploy
+steps:
+  main_clone:
+    title: Cloning main repository...
+    type: git-clone
+    repo: '${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}'
+    revision: '${{CF_REVISION}}'
+    stage: prepare
+    git: github-1
+  BuildProject:
+    title: Build project
+    stage: build
+    image: pulumi/pulumi
+    commands:
+      - yarn install
+  SelectMyCluster:
+    title: Select K8s cluster
+    stage: deploy
+    image: codefresh/kubectl:1.13.3
+    commands: 
+      - kubectl config get-contexts
+      - kubectl config use-context "kostis-demo@FirstKubernetes"
+  RunPulumi:
+    title: Deploying
+    stage: deploy
+    image: pulumi/pulumi
+    commands: 
+      - pulumi stack select dev --non-interactive
```

Comment:
I don't think the `--non-interactive` flags are required. If you run a `pulumi login`, and if the `PULUMI_ACCESS_TOKEN` env var is available, then a non-interactive login will take precedence anyway. It also looks like the login command is missing in your full example here?

Reply:
The non-interactive flag is certainly needed on the pulumi up command (I removed it from the other ones). If not provider, pulumi asks "do you want to perform this update, yes, no, details" and of course the CI/CD pipeline just blocks waiting for input.

Again I tried the example with and without an explicit pulumi login and it works both ways. But I added it as you suggested.

Reply:
Ah I see. You can provide an explicit approval for the update by passing the `--yes` flag to the `pulumi up` command. However, I am myself wondering if one is preferred over the other? Personally, I have been using the `--yes` flag since it reads better, that I am giving the CLI explicit approval to perform the update.

cc @lukehoban

Reply:
TBH, passing `--non-interactive` is safest. We auto-detect whether you're in CI based on well known environment variables and automatically disable interactive mode. So (a) it's possible we aren't detecting CodeFresh and (b) it's safer to just explicitly pass the flag anyway rather than depend on subtle implicit behavior. /cc @CyrusNajmabadi to check my understanding.